### ISSUE TITLE
Provide server-side validation

### ISSUE THREAD

**Issue by simongroenewolt** [bolt/core#1931]
At the moment there is no way to do server-side validation of content entered via the bolt editor. (There is a little bit of validation done via javascript in the browser.)

I'm proposing to make a minimal set of changes that will allow developers who are using bolt/core to build a site to specify their own validation in php. These changes also make it possible for bolt/core to include server side validation itself later, while allowing developers to override this.

How can you help?
-------------------------------

I'll be happy to discuss the options, and can provide an outline of the changes as I would see them.

I've started on a proof-of-concept, and can create a proposal for the change as a PR in the short future (days/weeks).

Outline of changes
-------------------------------

I think this would work:

The basic approach would be to use as much code and knowledge from [symfony/validator](https://github.com/symfony/validator) as possible, but to not force the use of it.

**Create a new interface definition**: `ContentValidatorInterface`, that has a single function: `validate()` - this function takes a content object, and returns an array-like structure of constraint violations. `ConstraintViolationListInterface` (used by `symfony/validator`) should definitely be allowed as a return value. The structure of the items in the array should follow the structure als used by `symfony/validator`.

**Add code to the `save()` function in `ContentEditController.php`** that will invoke the validate function if `ContentValidatorInterface` is available/injected. It should show the edit screen if the returned array has elements (instead of saving the data), and pass an additional 'errors' twig variable containing the violations.

**Change twig templates:** `templates/content/edit.html.twig`, and probably `templates/_partials/fields/_base.html.twig` templates to show the errors in the edit page.

With the above 3 changes a developer could create a validator class that implements `ContentValidatorInterface` as a service, and create as much validation as they want. Note that this is server-side only, implementing client-side validation that will match this should be left for another day.

**Comment by bobdenotter** [comment]
Hi, @simongroenewolt 👋 

Sounds like a good overall approach. Some initial thoughts: 

> The basic approach would be to use as much code and knowledge from symfony/validator as possible, but to not force the use of it.

Yes, agreed! We should "not re-invent the wheel", if possible in the slightest. 

Only thing I'm missing here is how the user (developer implementing a site/project) would trigger / configure this? Using a an extra attribute in `contenttypes.yaml`, like `validation: App\Project\MyValidators::email`? Event listener/subscriber? Perhaps even provide both, and leave it to the user?

**Comment by simongroenewolt** [comment]
@bobdenotter the way I'm envisioning the current proposal it that configuration is left to the developer using bolt - that is what makes this 'simple'.

All they have to do is create a service that implements ContentValidatorInterface. This service will be injected in the ContentEditController->save() call by Symfony and will be used to validate.
If you want to implement validation based on the configuration in contenttypes.yaml you can inject the Config or some other relevant class into your service and read settings from there -- that would be a sensible approach. 
From what I saw the $content object that is passed to the validate() function in the service contains all information necessary to create validation based on content types.
However, you can also create just a couple of if's based on the content type of the content you are validating for 1 or 2 specific cases, and leave it at that.

Now I'd think after adding this support, bolt should probably supply a default validator itself, or endorse a specific solution, and that is when your question about the configuration becomes relevant. I'd suggest following naming conventions and yaml notation of the symfony/validator package, but instead of putting the yaml in a separate validation.yaml file include it in the contenttypes.yaml via a property 'constraints'. After making the changes I'm suggesting, experimenting with the best solution for the configuration will be possible for any developer using bolt, as it's not needed to change code in bolt/core.
(But note again: this is not part of the current RFC)

I'll create a branch to show my ideas in code.

**Comment by simongroenewolt** [comment]
Small update: I now see that my initial mention about "to use as much code and knowledge from symfony/validator as possible" would imply using it in this proposal. However, I'm trying to make _this_ change without any dependencies on symfony/validator, and at the same time allow implementing a solution based on this change by using symfony/validator possible.

**Comment by simongroenewolt** [comment]
This is what I would change: https://github.com/bolt/core/compare/master...simongroenewolt:feature/server-side-validation-support

This branch has a (hand coded) validation on title length of Entries - they should be at least 10 long, otherwise you'll get a validation error.

Note: this implementation only shows the errors as one list at the top of the editor -- I think the only sane way to show the errors with their fields will be to handle the lookup in a twig extension. (I don't think we want to 'merge' the errors with the content object before passing it to twig, and I also don't think twig templating itself is suitable to build this lookup)

Note: the symfony validator uses path notation that [looks][like][this].

**Comment by simongroenewolt** [comment]
@bobdenotter hmm, I just discovered this version will make it possible to put constraints on fields, but relations, and taxonomy are updated as 'side effects' of the contentFromPost() call in ContentEditController and not available via the $content itself.

**Comment by bobdenotter** [comment]
> bolt should probably supply a default validator itself, or endorse a specific solution

Looking at this, i think so.. In fact, having this will allow us to make a default validator that can actually do these callbacks if defined in the contenttype. Two birds with one stone! 

>  I think the only sane way to show the errors with their fields will be to handle the lookup in a twig extension. 

I think we can expose them in twig to allow this. From a UX standpoint, it'd be best to show the errors twice: at the top, _and_ in context. But, that can be added later as well!

> hmm, I just discovered this version will make it possible to put constraints on fields, but relations, and taxonomy are updated as 'side effects' of the contentFromPost() call in ContentEditController and not available via the $content itself.

I think you're right. But, my approach with things like these is usually, that I don't feel it has to be perfect and complete at once. Developing a project like Bolt is very much an iterative process. 

Once this lands, Bolt will be a bit better. When people start using it, they might or might not require validation for relations and taxonomies. If so, we'll add them when we get there :-)

So, feel free to make a PR. We might want to tweak it a bit, but let's get the ball rolling!

**Comment by simongroenewolt** [comment]
> Looking at this, i think so.. In fact, having this will allow us to make a default validator that can actually do these callbacks if defined in the contenttype. Two birds with one stone!

Yes, that was on purpose :-)

I've added support for relations as well in my PR -- this is because I started this wanting to be able to use this for relations as well. (To 'fix' my issue with a non-optional relation rendering a select box and default-selecting the first relation available. I want that to render a select box with default nothing selected, but with a validation that will check the selection)

**Comment by simongroenewolt** [comment]
@bobdenotter What do you think would make sense for constraints on relations? 

We have these options:

1) Using the 'normal' constraints
```
        pages:
            multiple: true
            order: title
            label: Select one or more pages
            constraints:
                - Count:
                    min: 1
                    max: 50
```

2) Using a dedicated format, as the only constraint that I can envision to make sense is the count
```
        pages:
            multiple: true
            order: title
            label: Select one or more pages
            limit:
                min: 1
                max: 50
```
or even:
```
        pages:
            multiple: true
            order: title
            label: Select one or more pages
            min: 1
            max: 50
```

3) A third solution would be to encourage people to use the the dedicated format example, but allow the constraints option as well.

I think I'm leaning towards option 1 or 3. The 'problem' with suggesting to use the 'full' constraints option is that developers have to understand that they are validating an array of ids, and not a full list of related items.

**Comment by bobdenotter** [comment]
Option 1 would be nice for consistency, but this will bite us in the ass.. People see it, end they'll assume it works the same as constraints on Fields. Then they'll complain to us "Why can't I enforce the Relation to be a valid UUID?! Bolt sucks, i'm going back to Wordpress!" 

My vote would be for option 3. :-)

**Comment by I-Valchev** [comment]
I agree with both @bobdenotter  and @simongroenewolt about option 3. It would have been nice to keep the format _if they work the same_, but that's not an option right now. My vote goes to option 3 as well :-)

**Comment by simongroenewolt** [comment]
@bobdenotter @I-Valchev I've opted for option 2 at the moment and put the settings in a 'limit' node, this can be extended to allow option 3 in a later stage. If all works as expected, any developer _using_ bolt/core in a project will be able to extend/change the validation anyway -- so _technically_ we'd already have option 3, just not for website developers who don't want to create php code, which is fine I think.

**Comment by simongroenewolt** [comment]
Open issues:

- when validation fails the relations that were added/removed (but not saved) disappear (probably same for taxonomy)
- handling errors in the constraint configuration
- show errors next to fields in the editor
- scroll to error on form submit resulting in error? (Otherwise the user can assume the save was successful and just leave the page)

**Comment by simongroenewolt** [comment]
~~when validation fails the relations that were added/removed (but not saved) disappear (probably same for taxonomy)~~
done.

**Comment by simongroenewolt** [comment]
~~enable constraints on set and collection field types~~
done!

![Screen Shot 2020-10-19 at 23 43 35](https://user-images.githubusercontent.com/3960/96515076-f7e5e380-1264-11eb-946b-51ff97aeabd6.png)

**Comment by simongroenewolt** [comment]
@bobdenotter @I-Valchev I think the validator is (nearly?) feature-complete. I'd like to have some input on a couple of finishing touches.

I just did a push for the https://github.com/bolt/core/pull/2001 PR, some tests are failing at the moment because I've used an existing content type to add constraints to -- will fix that.

1) Handling errors in the constraint configuration in `contenttypes.yml`. Would this be necessary? What is the 'bolt way' of signaling errors in the config files?

2) Show errors next to fields in the editor. 

Bob wrote:
> I think we can expose them in twig to allow this. From a UX standpoint, it'd be best to show the errors twice: at the top, and in context. But, that can be added later as well!

Any suggestions? Code? 

3) It should be clear to the editor making the changes that saving failed, otherwise they might assume otherwise. I first thought a solution could be to scroll to the error(s) on top on form submit resulting in error? But maybe a modal(?) dialog with a message is better? This also a bit of a what would be the "bolt way" of doing this.

If item 3 is fixed it's in a usable state, but 1 and 2 would be very good for usability of course.
